rsocket/RSocketServiceHandler.h

```diff
@@ -0,0 +1,97 @@
+#include "rsocket/internal/Common.h"
+
+namespace rsocket {
+
+// This is a convenience function which is used to create a simplified
+// RSocketServiceHandler.
+using OnNewSetupFn =
+    folly::Function<std::shared_ptr<RSocketResponder>(const SetupParameters&)>;
+
+// This struct holds all the necessary information needed by the RSocketServer
+// to initiate a connection with a client.
+struct RSocketConnectionParams {
+  RSocketConnectionParams(
+      std::shared_ptr<RSocketResponder> _responder,
+      std::shared_ptr<RSocketStats> _stats = RSocketStats::noop(),
+      std::shared_ptr<RSocketNetworkStats> _networkStats = nullptr)
+      : responder(std::move(_responder)),
+        stats(std::move(_stats)),
+        networkStats(std::move(_networkStats)) {}
+  std::shared_ptr<RSocketResponder> responder;
+  std::shared_ptr<RSocketStats> stats;
+  std::shared_ptr<RSocketNetworkStats> networkStats;
+};
+
+// This class has to be implemented by the application.  The methods can be
+// called from different threads and it is the application's responsibility to
+// ensure thread-safety.
+class RSocketServiceHandler {
+ public:
+  virtual ~RSocketServiceHandler() = default;
+
+  // This method gets called for each client that connects to the server.  The
+  // application has to implement this method.  If the application does not
+  // want to accept this connection, it should return a RSocketException.  If
+  // the application returns a RSocketException, then an ERROR frame with code
+  // REJECTED_SETUP is sent to the client.  The exception message is sent as
+  // payload.
+  virtual folly::Expected<RSocketConnectionParams, RSocketException> onNewSetup(
+      const SetupParameters&) = 0;
+
+  // This method gets called after some state is created for each client.  The
+  // application should preserve the RSocketServerState if it wants to resume
+  // the connection later.
+  virtual void onNewRSocketState(
+      std::unique_ptr<RSocketServerState>,
+      ResumeIdentificationToken) {}
+
+  // This method gets called when a client tries to resume. The application
+  // should return the corresponding RSocketServerState for the given resumeId
+  // if it wants to resume.  The application should return a RSocketException
+  // if it does not want to resume.  If the application returns a
+  // RSocketException, then an ERROR frame with code REJECTED_RESUME is sent to
+  // the client.  The exception message is sent as payload.
+  virtual folly::Expected<std::unique_ptr<RSocketServerState>, RSocketException>
+  onResume(ResumeIdentificationToken) {
+    return nullptr;
+  }
+
+  // This method gives a fine-grained control to the application during
+  // resumption (provided it agreed to resume in the onResume() call above).
+  // It tells the application what streams will be resumed in clean/dirty
+  // state.  If the application is okay with resuming the given clean/dirty
+  // streams, it should return true.
+  virtual bool canResume(
```

Comment:
To have local-scope, the application should use per connection ServiceHandler.  If not it has to use the ResumeId to tie it back to a ServerState.